### ISSUE TITLE
Perl binary no longer relocatable

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#10910]
Migrated from [rt.perl.org#81026](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=81026) (status was 'resolved')

Searchable as RT81026$

**Comment by p5pRT** [comment]
### From @doughunt

#### Created by @doughunt



Hello&#8203;:  For several years now\, I have been building a set of relocatable
RPMs including perl and many perl modules\.  I have been using a trick
I learned from the ActiveState perl distribution&#8203;:  Compile perl with an
install PREFIX set to a long path name\.  Then when installing the RPM
\(in the RPM %post section\) one can 'relocate' the perl binary by swapping
the long PREFIX path with the shorter correct path and null padding the binary\.

In my perl RPM %build section\, I do something like this&#8203;:

sh Configure \-Dprefix=/tmp/LongPathnameForInstallBeforeRelocationToNewBase/ops/tools \-des

Then when I install the RPM \(in the %post section\) I run a perl script which
switches the install PREFIX \(/tmp/LongPathnameForInstallBeforeRelocationToNewBase/ops/tools\)
with the real install prefix \(say /ops/tools\) in the perl binary using NULLs to 
keep the binary the same length\.

This stopped working sometime in perl 5\.12\.  The use of the STR\_WITH\_LEN macro
has disabled run\-time checking of the length of embedded path names\.
For example&#8203;:

S\_incpush\_use\_sep\(aTHX\_ STR\_WITH\_LEN\(SITELIB\_EXP\)\, INCPUSH\_CAN\_RELOCATE\);

The fixes the length of the SITELIB library path to be the same as when compiled\,
thus ruining relocation\.

I've been using the patch below in production for a couple of months with
no ill effects\.  It passes the test suite\.
This patch just replaces the above line with

S\_incpush\_use\_sep\(aTHX\_ SITELIB\_EXP\, 0\, INCPUSH\_CAN\_RELOCATE\);

in several cases\.  The '0' means that the length of SITELIB\_EXP is 
computed at runtime with strlen\, allowing relocation to work\.

Please consider this patch\, I consider the ability to make relocatable RPMs of
perl to be an essential feature of the language that I and others have profited
from\.

Regards\, 

&nbsp;  Doug Hunt \(dhunt@&#8203;ucar\.edu\)


\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-%\<\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
<details><summary>Inline Patch</summary>

```diff
diff -r -u -N perl-5.12.2.orig/perl.c perl-5.12.2/perl.c
--- perl-5.12.2.orig/perl.c	2010-11-01 16:33:09.000000000 -0600
+++ perl-5.12.2/perl.c	2010-11-01 16:29:08.000000000 -0600
@@ -4121,7 +4121,7 @@
     SITEARCH, SITELIB, VENDORARCH, VENDORLIB, ARCHLIB and PRIVLIB
 */
 #ifdef APPLLIB_EXP
-    S_incpush_use_sep(aTHX_ STR_WITH_LEN(APPLLIB_EXP),
+    S_incpush_use_sep(aTHX_ APPLLIB_EXP, 0,
 		      INCPUSH_ADD_SUB_DIRS|INCPUSH_CAN_RELOCATE);
 #endif
 
@@ -4129,7 +4129,7 @@
     /* sitearch is always relative to sitelib on Windows for
      * DLL-based path intuition to work correctly */
 #  if !defined(WIN32)
-	S_incpush_use_sep(aTHX_ STR_WITH_LEN(SITEARCH_EXP),
+    S_incpush_use_sep(aTHX_ SITEARCH_EXP, 0,
 			  INCPUSH_CAN_RELOCATE);
 #  endif
 #endif
@@ -4141,7 +4141,7 @@
 	if (s)
 	    incpush_use_sep(s, len, INCPUSH_ADD_SUB_DIRS|INCPUSH_CAN_RELOCATE);
 #  else
-	S_incpush_use_sep(aTHX_ STR_WITH_LEN(SITELIB_EXP), INCPUSH_CAN_RELOCATE);
+	S_incpush_use_sep(aTHX_ SITELIB_EXP, 0, INCPUSH_CAN_RELOCATE);
 #  endif
 #endif
 
@@ -4149,7 +4149,7 @@
     /* vendorarch is always relative to vendorlib on Windows for
      * DLL-based path intuition to work correctly */
 #  if !defined(WIN32)
-    S_incpush_use_sep(aTHX_ STR_WITH_LEN(PERL_VENDORARCH_EXP),
+	S_incpush_use_sep(aTHX_ PERL_VENDORARCH_EXP, 0,
 		      INCPUSH_CAN_RELOCATE);
 #  endif
 #endif
@@ -4161,13 +4161,13 @@
 	if (s)
 	    incpush_use_sep(s, len, INCPUSH_ADD_SUB_DIRS|INCPUSH_CAN_RELOCATE);
 #  else
-	S_incpush_use_sep(aTHX_ STR_WITH_LEN(PERL_VENDORLIB_EXP),
+	S_incpush_use_sep(aTHX_ PERL_VENDORLIB_EXP, 0,
 			  INCPUSH_CAN_RELOCATE);
 #  endif
 #endif
 
 #ifdef ARCHLIB_EXP
-    S_incpush_use_sep(aTHX_ STR_WITH_LEN(ARCHLIB_EXP), INCPUSH_CAN_RELOCATE);
+	S_incpush_use_sep(aTHX_ ARCHLIB_EXP, 0, INCPUSH_CAN_RELOCATE);
 #endif
 
 #ifndef PRIVLIB_EXP
@@ -4182,12 +4182,12 @@
 #  ifdef NETWARE
     S_incpush_use_sep(aTHX_ PRIVLIB_EXP, 0, INCPUSH_CAN_RELOCATE);
 #  else
-    S_incpush_use_sep(aTHX_ STR_WITH_LEN(PRIVLIB_EXP), INCPUSH_CAN_RELOCATE);
+    S_incpush_use_sep(aTHX_ PRIVLIB_EXP, 0, INCPUSH_CAN_RELOCATE);
 #  endif
 #endif
 
 #ifdef PERL_OTHERLIBDIRS
-    S_incpush_use_sep(aTHX_ STR_WITH_LEN(PERL_OTHERLIBDIRS),
+    S_incpush_use_sep(aTHX_ PERL_OTHERLIBDIRS, 0,
 		      INCPUSH_ADD_VERSIONED_SUB_DIRS|INCPUSH_NOT_BASEDIR
 		      |INCPUSH_CAN_RELOCATE);
 #endif
@@ -4225,25 +4225,25 @@
     SITELIB and VENDORLIB for older versions
 */
 #ifdef APPLLIB_EXP
-    S_incpush_use_sep(aTHX_ STR_WITH_LEN(APPLLIB_EXP), INCPUSH_ADD_OLD_VERS
+    S_incpush_use_sep(aTHX_ APPLLIB_EXP, 0, INCPUSH_ADD_OLD_VERS
 		      |INCPUSH_NOT_BASEDIR|INCPUSH_CAN_RELOCATE);
 #endif
 
 #if defined(SITELIB_STEM) && defined(PERL_INC_VERSION_LIST)
     /* Search for version-specific dirs below here */
-    S_incpush_use_sep(aTHX_ STR_WITH_LEN(SITELIB_STEM),
+    S_incpush_use_sep(aTHX_ SITELIB_STEM, 0,
 		      INCPUSH_ADD_OLD_VERS|INCPUSH_CAN_RELOCATE);
 #endif
 
 
 #if defined(PERL_VENDORLIB_STEM) && defined(PERL_INC_VERSION_LIST)
     /* Search for version-specific dirs below here */
-    S_incpush_use_sep(aTHX_ STR_WITH_LEN(PERL_VENDORLIB_STEM),
+    S_incpush_use_sep(aTHX_ PERL_VENDORLIB_STEM, 0,
 		      INCPUSH_ADD_OLD_VERS|INCPUSH_CAN_RELOCATE);
 #endif
 
 #ifdef PERL_OTHERLIBDIRS
-    S_incpush_use_sep(aTHX_ STR_WITH_LEN(PERL_OTHERLIBDIRS),
+    S_incpush_use_sep(aTHX_ PERL_OTHERLIBDIRS, 0,
 		      INCPUSH_ADD_OLD_VERS|INCPUSH_ADD_ARCHONLY_SUB_DIRS
 		      |INCPUSH_CAN_RELOCATE);
 #endif
------------------%<------------------------------------------------
```
</details>


<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=high

Site configuration information for perl 5.12.2:

Configured by rpmbuild at Tue Dec  7 14:28:53 MST 2010.

Summary of my perl5 (revision 5 version 12 subversion 2) configuration:
   
  Platform:
    osname=linux, osvers=2.6.18-194.17.1.el5, archname=x86_64-linux
    uname='linux lynx.cosmic.ucar.edu 2.6.18-194.17.1.el5 #1 smp wed sep 29 12:50:31 edt 2010 x86_64 x86_64 x86_64 gnulinux '
    config_args='-Dprefix=/home/rpmbuild/rpm/aReallyLongNameForOps/tools -des'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.1.2 20080704 (Red Hat 4.1.2-48)', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib /lib64 /usr/lib64 /usr/local/lib64
    libs=-lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.5.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.5'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -L/usr/local/lib -fstack-protector'

Locally applied patches:
    


@INC for perl 5.12.2:
    /home/rpmbuild/rpm/aReallyLongNameForOps/tools/lib/perl5/site_perl/5.12.2/x86_64-linux
    /home/rpmbuild/rpm/aReallyLongNameForOps/tools/lib/perl5/site_perl/5.12.2
    /home/rpmbuild/rpm/aReallyLongNameForOps/tools/lib/perl5/5.12.2/x86_64-linux
    /home/rpmbuild/rpm/aReallyLongNameForOps/tools/lib/perl5/5.12.2
    .


Environment for perl 5.12.2:
    HOME=/home/rpmbuild
    LANG=C
    LANGUAGE (unset)
    LD_LIBRARY_PATH=/home/rpmbuild/rpm/aReallyLongNameForOps/tools/pgi/lib:/home/rpmbuild/rpm/aReallyLongNameForOps/tools/lib:/home/rpmbuild/rpm/aReallyLongNameForOps/tools/lib/qwt
    LOGDIR (unset)
    PATH=.:/home/rpmbuild/bin:/home/rpmbuild/rpm/aReallyLongNameForOps/tools/bin:/home/rpmbuild/rpm/aReallyLongNameForOps/tools/pgi/bin:/home/rpmbuild/rpm/aReallyLongNameForOps/tools/postgres/bin:/home/rpmbuild/rpm/aReallyLongNameForOps/tools/apache/bin:/home/rpmbuild/rpm/aReallyLongNameForOps/tools/bern5/BERN/PGM/EXE:/usr/local/dcs/bin:/usr/local/grace/bin:/home/rpmbuild/rpm/aReallyLongNameForOps/tools/ncarg/bin:/usr/lib64/qt-3.3/bin:/usr/kerberos/bin:/usr/local/bin:/bin:/usr/bin:/home/rpmbuild/bin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @jandubois

On Mon\, 20 Dec 2010\, Doug Hunt \(via RT\) wrote&#8203;:
> 
> This stopped working sometime in perl 5\.12\.  The use of the STR\_WITH\_LEN macro
> has disabled run\-time checking of the length of embedded path names\.
> For example&#8203;:
> 
> S\_incpush\_use\_sep\(aTHX\_ STR\_WITH\_LEN\(SITELIB\_EXP\)\, INCPUSH\_CAN\_RELOCATE\);
> 
> The fixes the length of the SITELIB library path to be the same as when compiled\,
> thus ruining relocation\.
> 
> I've been using the patch below in production for a couple of months with
> no ill effects\.  It passes the test suite\.
> This patch just replaces the above line with
> 
> S\_incpush\_use\_sep\(aTHX\_ SITELIB\_EXP\, 0\, INCPUSH\_CAN\_RELOCATE\);
> 
> in several cases\.  The '0' means that the length of SITELIB\_EXP is
> computed at runtime with strlen\, allowing relocation to work\.

ActivePerl 5\.12 uses the patch appended below to keep the relocation
mechanism working\. I'm find with S\_incpush\_use\_sep\(\) \*always\* letting
compute the length of the incoming string \(it only happens once\, so it
is not a worthwhile optimization\)\. If that is not acceptable\, then using
another preprocessor symbol in place of ACTIVEPERL\_VERSION in the patch
below would also allow people to re\-activate relocation support with a
Configure option instead of having to patch the sources\.

Cheers\,
\-Jan

commit e62d19888d3d571e861012dfd15d05237da4ee94
Author&#8203;: Gisle Aas \<gisle@&#8203;activestate\.com>
Date&#8203;:   Wed Feb 10 14&#8203;:29&#8203;:06 2010 \-0800

&nbsp;    For ActivePerl don't trust the string length passed to S\_incpush\_use\_sep
&nbsp;    
&nbsp;    We relocate the strings in the binary\, but the length determined by
&nbsp;    the STR\_WITH\_LEN macro is not adjusted accordingly\.

<details><summary>Inline Patch</summary>

```diff
diff --git a/perl.c b/perl.c
index 2659eb0..7aa03cc 100644
--- a/perl.c
+++ b/perl.c
@@ -4536,7 +4536,9 @@ S_incpush_use_sep(pTHX_ const char *p, STRLEN len, U32 flags)
 
     PERL_ARGS_ASSERT_INCPUSH_USE_SEP;
 
+#ifndef ACTIVEPERL_VERSION /* Can't trust STR_WITH_LEN after relocation */
     if (!len)
+#endif
        len = strlen(p);
 
     end = p + len;
```
</details>

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From zefram@fysh.org

This hack for relocation is fundamentally fragile\.  Off the top of my
head\, here are some other ways it could foreseeably break&#8203;:

\* we might bake precomputed hash values into the binary

\* the strings you want to munge can\, at the compiler's option\, share
&nbsp;  storage with other strings \("/foo/bar" can share storage with "/bar"\)
&nbsp;  or other constant data\, so your binary editing breaks other data

If we want to maintain the ability for this binary\-editing hack to work\,
with it being so easy to break by perfectly normal programming\, we will
at minimum need a test added to the test suite\, which checks whether it
still works\.

\-zefram

**Comment by p5pRT** [comment]
### From zefram@fysh.org

Jan Dubois wrote&#8203;:
>\+\#ifndef ACTIVEPERL\_VERSION /\* Can't trust STR\_WITH\_LEN after relocation \*/
>     if \(\!len\)
>\+\#endif
>        len = strlen\(p\);

Won't that break the processing of strings that are not nul terminated
or which contain embedded nuls?

\-zefram

**Comment by p5pRT** [comment]
### From @gisle

On Dec 21\, 2010\, at 12&#8203;:00 \, Zefram wrote&#8203;:

> Jan Dubois wrote&#8203;:
>> \+\#ifndef ACTIVEPERL\_VERSION /\* Can't trust STR\_WITH\_LEN after relocation \*/
>>   if \(\!len\)
>> \+\#endif
>>      len = strlen\(p\);
> 
> Won't that break the processing of strings that are not nul terminated
> or which contain embedded nuls?

No\, because S\_incpush\_use\_sep\(\) is not called with such strings\.  The p is a file name\.

\-\-Gisle

**Comment by p5pRT** [comment]
### From @gisle

On Dec 21\, 2010\, at 11&#8203;:59 \, Zefram wrote&#8203;:

> This hack for relocation is fundamentally fragile\.

That's true\.  It would have been better to end up with a binary with no
hardcoded prefix strings compiled into it\.  As long as we still have that
this relocation hack seems worth it\.

> Off the top of my head\, here are some other ways it could foreseeably break&#8203;:
> 
> \* we might bake precomputed hash values into the binary

We would have to patch around it if that happened\.

> \* the strings you want to munge can\, at the compiler's option\, share
> storage with other strings \("/foo/bar" can share storage with "/bar"\)
> or other constant data\, so your binary editing breaks other data

This was indeed a real problem for us\.  We compile ActivePerl with compiler options
to disable this optimization\.

> If we want to maintain the ability for this binary\-editing hack to work\,
> with it being so easy to break by perfectly normal programming\, we will
> at minimum need a test added to the test suite\, which checks whether it
> still works\.

ActivePerl is effectively such a test :\-\)

A test for this seems hard to implement in the perl test suite\.  You basically
would have to install perl with libraries to some location\, then relocate it a
shorter prefix and then run the relocated perl through the full test suite again\.

\-\-Gisle

**Comment by p5pRT** [comment]
### From @timj

On Mon\, Dec 20\, 2010 at 7&#8203;:50 AM\, Doug Hunt \<perlbug\-followup@&#8203;perl\.org>wrote&#8203;:

> \# New Ticket Created by  Doug Hunt
> \# Please include the string&#8203;:  \[perl \#81026\]
> \# in the subject line of all future correspondence about this issue\.
> \# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=81026 >
>
>
>
> This is a bug report for perl from dhunt@&#8203;ucar\.edu\,
> generated with the help of perlbug 1\.39 running under perl 5\.12\.2\.
>
>
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
>
> Hello&#8203;:  For several years now\, I have been building a set of relocatable
> RPMs including perl and many perl modules\.  I have been using a trick
> I learned from the ActiveState perl distribution&#8203;:  Compile perl with an
> install PREFIX set to a long path name\.  Then when installing the RPM
> \(in the RPM %post section\) one can 'relocate' the perl binary by swapping
> the long PREFIX path with the shorter correct path and null padding the
> binary\.
>
> In my perl RPM %build section\, I do something like this&#8203;:
>
> sh Configure
> \-Dprefix=/tmp/LongPathnameForInstallBeforeRelocationToNewBase/ops/tools \-des
>
>
How does all this relate to the \-Duserelocatableinc configure option? I
thought the "patch the binary after install" approach was deprecated when
that option was added\.

\-\- 
Tim Jenness

**Comment by p5pRT** [comment]
### From @jandubois

On Tue\, 21 Dec 2010\, Tim Jenness wrote&#8203;:
> How does all this relate to the \-Duserelocatableinc configure option? I
> thought the "patch the binary after install" approach was deprecated when
> that option was added\.

That code currently only works on Linux because it uses the /proc/self
filesystem to reliably get the location of the perl executable itself\.

The code isn't needed on Windows\, where Perl self\-relocates at runtime
since at least 5\.005\.

I know how to make \-Duserelocatableinc work for OS X and Solaris\, but
for other platforms it may be hard/impossible to reliably find the
full path\.  Looking for argv\[0\] on the PATH opens up all manner of
security vulnerabilities\, which is the reason for the current system
of hardcoded locations for things like the privlib directory\.

Cheers\,
\-Jan

**Comment by p5pRT** [comment]
### From @doughunt

Jan&#8203;:  I like your patch better than mine and agree that a preprocessor
symbol other than ACTIVEPERL\_VERSION would be best\.

Regards\, 

&nbsp;  Doug

On Mon Dec 20 23&#8203;:41&#8203;:54 2010\, jdb wrote&#8203;:
> On Mon\, 20 Dec 2010\, Doug Hunt \(via RT\) wrote&#8203;:
> >
> > This stopped working sometime in perl 5\.12\.  The use of the
> STR\_WITH\_LEN macro
> > has disabled run\-time checking of the length of embedded path names\.
> > For example&#8203;:
> >
> > S\_incpush\_use\_sep\(aTHX\_ STR\_WITH\_LEN\(SITELIB\_EXP\)\,
> INCPUSH\_CAN\_RELOCATE\);
> >
> > The fixes the length of the SITELIB library path to be the same as
> when compiled\,
> > thus ruining relocation\.
> >
> > I've been using the patch below in production for a couple of months
> with
> > no ill effects\.  It passes the test suite\.
> > This patch just replaces the above line with
> >
> > S\_incpush\_use\_sep\(aTHX\_ SITELIB\_EXP\, 0\, INCPUSH\_CAN\_RELOCATE\);
> >
> > in several cases\.  The '0' means that the length of SITELIB\_EXP is
> > computed at runtime with strlen\, allowing relocation to work\.
> 
> ActivePerl 5\.12 uses the patch appended below to keep the relocation
> mechanism working\. I'm find with S\_incpush\_use\_sep\(\) \*always\* letting
> compute the length of the incoming string \(it only happens once\, so it
> is not a worthwhile optimization\)\. If that is not acceptable\, then
> using
> another preprocessor symbol in place of ACTIVEPERL\_VERSION in the
> patch
> below would also allow people to re\-activate relocation support with a
> Configure option instead of having to patch the sources\.
> 
> Cheers\,
> \-Jan
> 
> commit e62d19888d3d571e861012dfd15d05237da4ee94
> Author&#8203;: Gisle Aas \<gisle@&#8203;activestate\.com>
> Date&#8203;:   Wed Feb 10 14&#8203;:29&#8203;:06 2010 \-0800
> 
>     For ActivePerl don't trust the string length passed to
> S\_incpush\_use\_sep
> 
>     We relocate the strings in the binary\, but the length determined
> by
>     the STR\_WITH\_LEN macro is not adjusted accordingly\.
> 
> diff \-\-git a/perl\.c b/perl\.c
> index 2659eb0\.\.7aa03cc 100644
> \-\-\- a/perl\.c
> \+\+\+ b/perl\.c
> @&#8203;@&#8203; \-4536\,7 \+4536\,9 @&#8203;@&#8203; S\_incpush\_use\_sep\(pTHX\_ const char \*p\, STRLEN
> len\, U32 flags\)
> 
>      PERL\_ARGS\_ASSERT\_INCPUSH\_USE\_SEP;
> 
> \+\#ifndef ACTIVEPERL\_VERSION /\* Can't trust STR\_WITH\_LEN after
> relocation \*/
>      if \(\!len\)
> \+\#endif
>         len = strlen\(p\);
> 
>      end = p \+ len;
>

**Comment by p5pRT** [comment]
### From @pjcj

On Tue\, Dec 21\, 2010 at 11&#8203;:44&#8203;:38AM \-0800\, Jan Dubois wrote&#8203;:

> I know how to make \-Duserelocatableinc work for \[\.\.\.\] Solaris

The last time I looked at this I ended up with the following patch which
sorted thing out for Solaris but broke linux\.

 http&#8203;://www\.nntp\.perl\.org/group/perl\.perl5\.porters/2009/03/msg145214\.html

The was some funny stuff going on somewhere and I never managed to find
the time to investigate further\, and now I don't have access to a
Solaris box\.  If you have a proper fix that would be great\.

\-\- 
Paul Johnson \- paul@&#8203;pjcj\.net
http&#8203;://www\.pjcj\.net

**Comment by p5pRT** [comment]
### From @timj


On Dec 21\, 2010\, at 9&#8203;:44 AM\, Jan Dubois wrote&#8203;:

> On Tue\, 21 Dec 2010\, Tim Jenness wrote&#8203;:
>> How does all this relate to the \-Duserelocatableinc configure option? I
>> thought the "patch the binary after install" approach was deprecated when
>> that option was added\.
> 
> That code currently only works on Linux because it uses the /proc/self
> filesystem to reliably get the location of the perl executable itself\.
> 
> The code isn't needed on Windows\, where Perl self\-relocates at runtime
> since at least 5\.005\.
> 
> I know how to make \-Duserelocatableinc work for OS X and Solaris\, but

I dream of this working right on OSX\. I had tried it once and got some test failures but there was nothing in the INSTALL notes to indicate the userelocatableinc was linux only\.

\-\- 
Tim Jenness

**Comment by p5pRT** [comment]
### From @doughunt

Hi Jan&#8203;:  This sounds like a good idea&#8203;:  use your patch but change the 
preprocessor symbol from ACTIVEPERL\_VERSION to RELOCATABLE\_VERSION or some 
such\.  I like your patch better anyway\!

Regards\,

&nbsp;   Doug Hunt

dhunt@&#8203;ucar\.edu
Software Engineer
UCAR \- COSMIC\, Tel\. \(303\) 497\-2611

On Mon\, 20 Dec 2010\, Jan Dubois via RT wrote&#8203;:

> On Mon\, 20 Dec 2010\, Doug Hunt \(via RT\) wrote&#8203;:
>>
>> This stopped working sometime in perl 5\.12\.  The use of the STR\_WITH\_LEN macro
>> has disabled run\-time checking of the length of embedded path names\.
>> For example&#8203;:
>>
>> S\_incpush\_use\_sep\(aTHX\_ STR\_WITH\_LEN\(SITELIB\_EXP\)\, INCPUSH\_CAN\_RELOCATE\);
>>
>> The fixes the length of the SITELIB library path to be the same as when compiled\,
>> thus ruining relocation\.
>>
>> I've been using the patch below in production for a couple of months with
>> no ill effects\.  It passes the test suite\.
>> This patch just replaces the above line with
>>
>> S\_incpush\_use\_sep\(aTHX\_ SITELIB\_EXP\, 0\, INCPUSH\_CAN\_RELOCATE\);
>>
>> in several cases\.  The '0' means that the length of SITELIB\_EXP is
>> computed at runtime with strlen\, allowing relocation to work\.
>
> ActivePerl 5\.12 uses the patch appended below to keep the relocation
> mechanism working\. I'm find with S\_incpush\_use\_sep\(\) \*always\* letting
> compute the length of the incoming string \(it only happens once\, so it
> is not a worthwhile optimization\)\. If that is not acceptable\, then using
> another preprocessor symbol in place of ACTIVEPERL\_VERSION in the patch
> below would also allow people to re\-activate relocation support with a
> Configure option instead of having to patch the sources\.
>
> Cheers\,
> \-Jan
>
> commit e62d19888d3d571e861012dfd15d05237da4ee94
> Author&#8203;: Gisle Aas \<gisle@&#8203;activestate\.com>
> Date&#8203;:   Wed Feb 10 14&#8203;:29&#8203;:06 2010 \-0800
>
>    For ActivePerl don't trust the string length passed to S\_incpush\_use\_sep
>
>    We relocate the strings in the binary\, but the length determined by
>    the STR\_WITH\_LEN macro is not adjusted accordingly\.
>
> diff \-\-git a/perl\.c b/perl\.c
> index 2659eb0\.\.7aa03cc 100644
> \-\-\- a/perl\.c
> \+\+\+ b/perl\.c
> @&#8203;@&#8203; \-4536\,7 \+4536\,9 @&#8203;@&#8203; S\_incpush\_use\_sep\(pTHX\_ const char \*p\, STRLEN len\, U32 flags\)
>
>     PERL\_ARGS\_ASSERT\_INCPUSH\_USE\_SEP;
>
> \+\#ifndef ACTIVEPERL\_VERSION /\* Can't trust STR\_WITH\_LEN after relocation \*/
>     if \(\!len\)
> \+\#endif
>        len = strlen\(p\);
>
>     end = p \+ len;
>
>

**Comment by p5pRT** [comment]
### From @jandubois

Fixed in http&#8203;://perl5\.git\.perl\.org/perl\.git/commitdiff/f31c6eed

**Comment by p5pRT** [comment]
@jandubois - Status changed from 'open' to 'resolved'
